File: javascript/src/providers/Emailfake.js
import { TempMailProvider } from '../base.js';
import { Message, MessageDetail } from '../models.js';
import { TempMailError, NotFoundError } from '../errors.js';

const CHARS = 'abcdefghijklmnopqrstuvwxyz0123456789';

/**
 * emailfake.com provider — surl cookie + channel URL, HTML scraping.
 *
 * This is the base for the emailfake / generator.email / email-temp family:
 * they share the same backend (gasmurl), the same `surl` cookie, the same
 * `email-table` inbox markup and `from_div_*`/`subj_div_*`/`time_div_*` fields.
 * Subclasses only override the base URL and provider name.
 *
 * HTML is parsed with regex — no DOMParser in Node, no deps added.
 */
export class Emailfake extends TempMailProvider {
  /**
   * @param {{ randomUA?: boolean, useCookies?: boolean }|undefined} opts
   * @param {string} [name]
   * @param {string} [base]
   */
  constructor(opts = {}, name = 'emailfake', base = 'https://emailfake.com') {
    super(name, { randomUA: opts.randomUA, useCookies: true });
    this._base = base;
    /** @type {string|null} */
    this._email = null;
    /** @type {string|null} */
    this._domain = null;
    /** @type {string|null} */
    this._username = null;
  }

  /**
   * Fetch a /channel{1-9}/ page and scrape domain options from it.
   * @returns {Promise<string[]>}
   */
  async _getDomains() {
    const ch = 1 + Math.floor(Math.random() * 9);
    const res = await this._fetch(`${this._base}/channel${ch}/`);
    const html = await res.text();
    const domains = new Set();
    const re = /<option[^>]*value="([^"]+)"/gi;
    let m;
    while ((m = re.exec(html))) {
      const v = m[1].trim();
      if (v.includes('.') && !v.includes(' ') && !v.includes('@')) domains.add(v);
    }
    if (!domains.size) {
      // Fallback: look for domain-like strings in text content
      const textRe = />([a-z0-9-]+\.[a-z.]{2,})</gi;
      let m2;
      while ((m2 = textRe.exec(html))) {
        const v = m2[1].trim();
        if (!v.includes('emailfake') && !v.includes('generator') && !v.includes('email-temp')) {
          domains.add(v);
        }
      }
    }
    if (!domains.size) throw new TempMailError('No domains found on page', this.name);
    return [...domains];
  }

  /** @returns {Promise<string>} */
  async generateEmail() {
    const domains = await this._getDomains();
    this._domain = domains[Math.floor(Math.random() * domains.length)];
    this._username = Array.from({ length: 10 }, () =>
      CHARS[Math.floor(Math.random() * CHARS.length)],
    ).join('');
    this._email = `${this._username}@${this._domain}`;
    // surl cookie routes the channel page to this mailbox: {domain}/{username}
    this.http.cookieJar['surl'] = `${this._domain}/${this._username}`;
    return this._email;
  }

  /**
   * @param {string} _email — unused; the surl cookie carries the mailbox
   * @returns {Promise<Message[]>}
   */
  async getInbox(_email) {
    if (!this._domain || !this._username) {
      throw new TempMailError('No email generated — call generateEmail() first', this.name);
    }
    const ch = 1 + Math.floor(Math.random() * 9);
    const res = await this._fetch(`${this._base}/channel${ch}/`);
    const html = await res.text();
    const table = this._extractBlockById(html, 'email-table');
    if (!table) return [];
    const messages = [];
    const linkRe = /<a[^>]*class="[^"]*list-group-item[^"]*"[^>]*>([\s\S]*?)<\/a>/gi;
    let lm;
    while ((lm = linkRe.exec(table))) {
      const block = lm[0];
      const href = (block.match(/href="([^"]+)"/) || [])[1] || '';
      const msgId = href ? href.replace(/\/+$/, '').split('/').pop() : '';
      if (!msgId || msgId.length < 10) continue;
      const sender = this._divText(block, 'from');
      const subject = this._divText(block, 'subj');
      const time = this._divText(block, 'time');
      messages.push(new Message({ id: msgId, sender, subject, date: time }));
    }
    return messages;
  }

  /**
   * @param {string} messageId
   * @returns {Promise<MessageDetail>}
   */
  async readMessage(messageId) {
    if (!this._domain || !this._username) {
      throw new TempMailError('No email generated — call generateEmail() first', this.name);
    }
    const url = `${this._base}/${this._domain}/${this._username}/${messageId}`;
    const res = await this._fetch(url);
    const html = await res.text();
    const body = this._extractBlockById(html, 'message');
    if (!body) throw new NotFoundError(`Message ${messageId} body not found`, this.name);
    return new MessageDetail({
      id: messageId,
      sender: this._divText(html, 'from_div'),
      subject: this._divText(html, 'subj_div'),
      date: this._divText(html, 'time_div'),
      bodyText: this._stripHtml(body),
      bodyHtml: body,
      attachments: [],
    });
  }

  /**
   * @param {string} _email
   * @returns {Promise<boolean>}
   */
  async deleteEmail(_email) {
    this._email = null;
    this._domain = null;
    this._username = null;
    return true;
  }

  // --- regex HTML helpers (no DOMParser in Node, no deps) ---

  /**
   * Extract the inner HTML of `<div id="{id}">` by counting balanced div tags.
   * Returns null if the opening tag is absent; falls back to the rest of the
   * document if no matching close is found.
   * @param {string} html
   * @param {string} id
   * @returns {string|null}
   */
  _extractBlockById(html, id) {
    const open = html.match(new RegExp(`<div[^>]*id="${id}"[^>]*>`, 'i'));
    if (!open) return null;
    const start = open.index + open[0].length;
    let depth = 1;
    const tagRe = /<\/?div\b[^>]*>/gi;
    tagRe.lastIndex = start;
    let t;
    while ((t = tagRe.exec(html))) {
      depth += t[0].startsWith('</') ? -1 : 1;
      if (depth === 0) return html.slice(start, t.index);
    }
    return html.slice(start);
  }

  /**
   * Text content of the first `<div class="…{classFrag}…">…</div>` (non-greedy).
   * @param {string} html
   * @param {string} classFrag
   * @returns {string}
   */
  _divText(html, classFrag) {
    const re = new RegExp(
      `<div[^>]*class="[^"]*${classFrag}[^"]*"[^>]*>([\\s\\S]*?)<\\/div>`,
      'i',
    );
    const m = html.match(re);
    return m ? this._stripHtml(m[1]) : '';
  }

  /**
   * Strip HTML tags and decode common entities.
   * @param {string} html
   * @returns {string}
   */
  _stripHtml(html) {
    if (!html) return '';
    return html
      .replace(/<[^>]*>/g, '')
      .replace(/&nbsp;/g, ' ')
      .replace(/&amp;/g, '&')
      .replace(/&lt;/g, '<')
      .replace(/&gt;/g, '>')
      .replace(/&quot;/g, '"')
      .trim();
  }
}

File: python/tempmail_wrapper/providers/emailfake.py
"""emailfake.com provider implementation (HTML scraping with surl cookie)."""

from __future__ import annotations

import random
import re
import string
from datetime import datetime

from bs4 import BeautifulSoup

from ..base import TempMailProvider
from ..exceptions import NotFoundError, TempMailError
from ..http_client import HttpClient
from ..models import Message, MessageDetail


class EmailfakeProvider(TempMailProvider):
    """Provider for emailfake.com — surl cookie + channel URL, HTML scraping."""

    BASE_URL = "https://emailfake.com"
    COOKIE_DOMAIN = ".emailfake.com"

    def __init__(
        self,
        proxies: list[str] | None = None,
        random_ua: bool = True,
        **kwargs,
    ) -> None:
        self._http = HttpClient(proxies=proxies, random_ua=random_ua, use_cookies=True)
        self._email: str | None = None
        self._domain: str | None = None
        self._username: str | None = None

    @property
    def name(self) -> str:
        return "emailfake"

    def _get_domains(self) -> list[str]:
        """Fetch available domains from the page."""
        import random as _r
        ch = _r.randint(1, 9)
        resp = self._http.get(f"{self.BASE_URL}/channel{ch}/", timeout=30)
        if resp.status_code != 200:
            raise TempMailError(f"Failed to load page: {resp.status_code}", self.name)
        soup = BeautifulSoup(resp.text, "html.parser")
        # Domains are in option tags or in the page text after @
        domains = []
        for opt in soup.find_all("option"):
            val = opt.get("value", "").strip()
            if "." in val and " " not in val and "@" not in val:
                domains.append(val)
        if not domains:
            # Fallback: parse from text - look for domain-like strings
            for text in soup.stripped_strings:
                if re.match(r"^[a-z0-9-]+\.[a-z.]{2,}$", text) and "emailfake" not in text:
                    domains.append(text)
        if not domains:
            raise TempMailError("No domains found on page", self.name)
        return list(set(domains))

    def generate_email(self) -> str:
        domains = self._get_domains()
        self._domain = random.choice(domains)
        self._username = "".join(
            random.choices(string.ascii_lowercase + string.digits, k=10)
        )
        self._email = f"{self._username}@{self._domain}"
        # Set surl cookie: surl={domain}/{username}
        self._http.session.cookies.set(
            "surl", f"{self._domain}/{self._username}", domain=self.COOKIE_DOMAIN
        )
        return self._email

    def get_inbox(self, email: str) -> list[Message]:
        if not self._domain or not self._username:
            raise TempMailError("No email generated — call generate_email() first", self.name)
        ch = random.randint(1, 9)
        resp = self._http.get(f"{self.BASE_URL}/channel{ch}/", timeout=30)
        if resp.status_code != 200:
            raise TempMailError(f"Failed to load inbox: {resp.status_code}", self.name)
        soup = BeautifulSoup(resp.text, "html.parser")
        email_table = soup.find("div", id="email-table")
        if not email_table:
            return []
        messages = []
        for item in email_table.find_all("div", class_=re.compile(r"list-group-item")):
            # Skip the outer container if it has nested list-group-items
            if item.find("div", class_=re.compile(r"list-group-item")):
                continue
            
            from_div = item.find("div", class_=re.compile(r"from_div"))
            subj_div = item.find("div", class_=re.compile(r"subj_div"))
            time_div = item.find("div", class_=re.compile(r"time_div"))
            
            sender = from_div.get_text(strip=True) if from_div else ""
            subject = subj_div.get_text(strip=True) if subj_div else ""
            time_str = time_div.get_text(strip=True) if time_div else ""
            
            # Use subject+time as message ID since there's no hash in the list view
            msg_id = f"{subject}_{time_str}".replace(" ", "_")
            
            try:
                date = datetime.strptime(time_str, "%Y-%m-%d %H:%M:%S")
            except ValueError:
                date = datetime.now()
            
            if sender or subject:  # Only add if we found actual content
                messages.append(
                    Message(id=msg_id, sender=sender, subject=subject, date=date)
                )
        return messages

    def read_message(self, message_id: str) -> MessageDetail:
        if not self._domain or not self._username:
            raise TempMailError("No email generated — call generate_email() first", self.name)
        url = f"{self.BASE_URL}/{self._domain}/{self._username}/{message_id}"
        resp = self._http.get(url, timeout=30)
        if resp.status_code != 200:
            raise NotFoundError(f"Message {message_id} not found", self.name)
        soup = BeautifulSoup(resp.text, "html.parser")
        msg_div = soup.find("div", id="message")
        if not msg_div:
            raise NotFoundError(f"Message {message_id} body not found", self.name)
        body_html = str(msg_div)
        body_text = msg_div.get_text(separator="\n", strip=True)
        # Try to extract sender/subject from the page
        from_div = soup.find("div", class_=re.compile(r"from_div"))
        subj_div = soup.find("div", class_=re.compile(r"subj_div"))
        time_div = soup.find("div", class_=re.compile(r"time_div"))
        sender = from_div.get_text(strip=True) if from_div else ""
        subject = subj_div.get_text(strip=True) if subj_div else ""
        time_str = time_div.get_text(strip=True) if time_div else ""
        try:
            date = datetime.strptime(time_str, "%Y-%m-%d %H:%M:%S")
        except ValueError:
            date = datetime.now()
        return MessageDetail(
            id=message_id,
            sender=sender,
            subject=subject,
            date=date,
            body_text=body_text,
            body_html=body_html,
            attachments=[],
        )

    def delete_email(self, email: str) -> bool:
        self._email = None
        self._domain = None
        self._username = None
        return True
